Title: About
Date: 2017-06-26 11:00
Category: About
 
## Our Mission
The Informatics and Scientific Applications group is a small group working in the cramped space between faculty, core facilities and high performance computing. We have a number of core responsibilities including first pass analysis for the sequencing core and software/analysis support for the proteomics mass-spectrometry core.     We also support the infrastructure for instrument scheduling and invoicing across many FAS facilities.

A large part of our remit is to smooth the path of scientific discovery by working with faculty, post-docs and grad students to design and analyze sequencing experiments for the most efficient bang for the buck.  Our group is currently small (two bioinformaticians and two web developers) so we have limited resources to fulfill this area but we are aiming to grow and expand our services.

## Services

The Informatics group provides bioinformatics analysis support for all users of the Core Sequencing facility and to FAS faculty groups regardless of the origin of their data.    In addition we have the responsibility for sample submission, tracking, scheduling and billing services for many of the core instrument facilities across FAS.

We are small in number and try to cover the needs of as many people across campus as possible.  This means, however, that we can’t provide in-depth or long-term analysis services for single groups.   In these cases we encourage PIs to budget for a bioinformatics post-doc in their grant applications.   We are very happy to provide letters of support and text detailing our expertise,  and outlining possible analyses in grant applications.   Additionally Scott Yockel in Research Computing is happy to provide letters of support for computational hardware.


## Team

### Tim Sackton
<div>
    <div class="team-photo">
        <figure>
            <a class="img" href="/images/tim.jpg">
                    <img class="img-responsive" src="/images/tim.jpg"></img>
            </a>
            <figcaption></figcaption>
        </figure>
    </div>
    <div class="team-blurb">
        <p>
        Director of Bioinformatics
        </p>
        <p>
        <a href="mailto:tsackton@g.harvard.edu">tsackton at harvard</a>
        </p>
    </div>
    <div style="clear: both;"></div>
</div>

### Aaron Kitzmiller
<div>
      <div class="team-photo">
                <figure>
                        <a class="img" href="/images/aaron.jpg">
                                <img class="img-responsive" src="/images/aaron.jpg"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
                Director of Software Operations
                </p>
                <p>
                 <a href="mailto:aaron_kitzmiller@harvard.edu">aaron_kitzmiller at harvard</a>
                </p>
        </div>
        <div style="clear: both;"></div>
</div>

### Meghan Porter-Mahoney
<div>
        <div class="team-photo">
                <figure>
                        <a class="img" href="/images/meghan.png">
                          <img class="img-responsive" src="/images/meghan.png"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
                Senior Software Engineer BA in Biochemistry and Molecular Biology, Boston University, MA in Information Technology Systems Management, Harvard University Extension School. Careers previous to software include: Molecular Biologist, Peace Corps Volunteer.
                </p>
                <p>
                <a href="mailto:mportermahoney@g.harvard.edu">mportermahoney at harvard</a>
                </p>
        </div>
        <div style="clear:both;"></div>
</div>

### Adam Freedman
<div>
        <div class="team-photo">
                <figure>
                        <a class="img" href="/images/adam_freedman.png">
                          <img class="img-responsive" src="/images/adam_freedman.png"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
                Data Scientist maintaining pipeline for Illumina sequencer data processing, also working on methods for analysis of RNA-seq (and other omics) data.</p>
                <p>
                <a href="mailto:adamfreedman@fas.harvard.edu">adamfreedman at harvard</a>
                </p>
        </div>
        <div style="clear:both;"></div>
</div>

### John Gaspar
<div>
        <div class="team-photo">
                <figure>
                        <a class="img" href="/images/john_gaspar.jpg">
                          <img class="img-responsive" src="/images/john_gaspar.jpg"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
                Bioinformatics Scientist, specializing in data processing and genomic enrichment assays.</p>
                <p>
                <a href="mailto:jgaspar@fas.harvard.edu">jgaspar at harvard</a>
                </p>
        </div>
        <div style="clear:both;"></div>
</div>

### Yasin Kaymaz
<div>
        <div class="team-photo">
                <figure>
                        <a class="img" href="/images/yasin_kaymaz.jpg">
                          <img class="img-responsive" src="/images/yasin_kaymaz.jpg"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
                Bioinformatics Scientist working in collaborative projects with Boehringer Ingelheim on brain circuitry underlying symptom domains of relevance to the central nervous system. These projects involve various types of single cell Next generation sequencing data analysis, integration, and interpretation. 
He gained his B.Sc. in Biochemistry (Ege University), and M.Sc. in Bioinformatics from Boston University. He graduated from University of Massachusetts Medical school with a PhD degree in Bioinformatics and Computational Biology.</p>
                <p>
                <a href="mailto:yasinkaymaz@fas.harvard.edu">yasinkaymaz at harvard</a>
                </p>
        </div>
        <div style="clear:both;"></div>
</div>

### Tommy Tang
<div>
        <div class="team-photo">
                <figure>
                        <a class="img" href="/images/tommy_pic.jpeg">
                          <img class="img-responsive" src="/images/tommy_pic.jpeg"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
Bioinformatics scientist working closely with Catherine Dulac's lab to catalog and understand the diversity and function of cell types in the mouse brain using single-cell RNA-seq and other cutting-edge techniques. He gained his PhD in Genetics and Genomics at the University of Florida and did a postdoc in computational biology at MD Anderson Cancer Center. He cares about reproducible research and open science.</p>
                <p>
                <a href="mailto:mingtang@fas.harvard.edu">mingtang at harvard</a>
                </p>
        </div>
        <div style="clear:both;"></div>
</div>

### Juerg Straubhaar
<div>
        <div class="team-photo">
                <figure>
                        <a class="img" href="/images/placeholder-profile-pic-large.png">
                          <img class="img-responsive" src="/images/placeholder-profile-pic-large.png"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
Bioinformatics scientist working closely with Doug Melton's lab on single-cell RNA-seq and other methods. Also does proteomics.</p>
                <p>
                <a href="mailto:jstraubhaar@fas.harvard.edu">jstraubhaar at harvard</a>
                </p>
        </div>
        <div style="clear:both;"></div>
</div>

### Brian Arnold
<div>
        <div class="team-photo">
                <figure>
                        <a class="img" href="/images/placeholder-profile-pic-large.png">
                          <img class="img-responsive" src="/images/placeholder-profile-pic-large.png"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
Bioinformatics scientist working on population genetics across a diverse taxonomic range including bacteria, plants, and animals.</p>
                <p>
                <a href="mailto:barnold@hsph.harvard.edu">barnold at harvard</a>
                </p>
        </div>
        <div style="clear:both;"></div>
</div>



## Alumni

### Michele Clamp

<div>
    <div class="team-photo">
        <figure>
            <a class="img" href="/images/michele.jpg">
                    <img class="img-responsive" src="/images/michele.jpg"></img>
            </a>
            <figcaption></figcaption>
        </figure>
    </div>
    <div class="team-blurb">
        <p>
        Former Director of Informatics and Scientific Applications   Yes those are aggressive filters on the photo – what of it?
        </p>
    </div>
    <div style="clear: both;"></div>
</div>

### Allison Shultz
<div>
        <div class="team-photo">
                <figure>
                        <a class="img" href="/images/allison_shultz.jpg">
                          <img class="img-responsive" src="/images/allison_shultz.jpg"></img>
                        </a>
                    <figcaption></figcaption>
                </figure>
        </div>
        <div class="team-blurb">
                <p>
                Former Bioinformatics Scientist working on analyses of genomic data, specializing on population genomics. Currently Curator of Birds, LA County Museum. Learn more about her research on her <a href="http://www.allisonshultz.com">website</a>.</p>
                <p>
                <a href="mailto:ashultz@fas.harvard.edu">ashultz at harvard</a>
                </p>
        </div>
        <div style="clear:both;"></div>

</div>


## Where to find us
Informatics and Scientific Applications is located on the first floor of 38 Oxford Street, Cambridge, MA 02138. Please note that this building is not actually on Oxford Street. It is located just to the east (or behind) the Northwest Labs, and just north of the Biological Labs. Click [here](http://map.harvard.edu/?ctrx=760110&amp;ctry=2963613&amp;level=8&amp;layers=Campus%20Base%20and%20Buildings,Map%20Text&amp;bld=04350) for an interactive map and click [here](https://maps.google.com/maps?q=38+Oxford+St,+Cambridge,+MA&amp;hl=en&amp;ll=42.379391,-71.113311&amp;spn=0.006221,0.011297&amp;sll=42.37839,-71.114729&amp;sspn=0.099544,0.180759&amp;oq=38+&amp;hnear=38+Oxford+St,+Cambridge,+Massachusetts+02138&amp;t=m&amp;z=17) for a Google map.

Use the entrance on the southwest corner of the building (marked in the map below), which is unlocked during regular business hours. Climb the ramp and go through the doors, then follow the hallway to our communal meeting area.

If you are planning to drive, please be advised that on-street parking is very limited and University-owned lots require a pre-paid day [permit](https://www2.uos.harvard.edu/cgi-bin/permit/purchase.pl). We strongly urge using public transportation.

![Map](/images/map.jpg)


###Directions
**By Public Transportation**
Take the Red Line to the Harvard Square stop, and choose the “Church Street” exit. Exit through the turnstiles and up the stairway on your left, which leads to Harvard Yard. Walk through the gate into Harvard Yard and proceed straight ahead to the statue of John Harvard. From here take the path to your left, exit the Yard, and veer slightly to the right. The Science Center (a large, concrete structure with a rock formation in front) will be on your left, and Memorial Hall (a large, ornate red brick building with a tall steeple) on your right. This path will take you to the intersection of Kirkland and Oxford Streets. Cross the street, and proceed left down Oxford Street. You will see the Harvard Museum of Natural History on your right. Take the first path on the right after the Museum and follow to the [main entrance of the Northwest Building](http://rc.fas.harvard.edu/wp-content/uploads/2011/09/Northwest-Building.jpg). When you get to the entrance of the Northwest Building turn right and follow the path that leads to the back of the building. When you get to the end of this path, the 38 Oxford Street building will be in front of you.

**From the North**
Take I-93 or Route 1 south toward Boston. Take the exit for “Storrow Drive-Cambridge.” Follow Storrow Drive west for approximately 4 miles. After approximately 2 miles Storrow Drive becomes Soldiers Field road. You will not notice any change. Stay in the right hand lane in order to avoid the many left hand exits off Storrow Drive until you reach the exit marked “Harvard Square/Cambridge.” Take the exit and go over the Charles River to the set of lights at Memorial Drive. Go straight ahead through the lights. You will now be on JFK Street. After you pass the intersection of JFK and Mt. Auburn Street, you will come to another set of lights where Nini’s Corner and The Coop are on the left, and Out of Town News and the subway “T” stop will be on your right. Go straight (veering slightly to the right) through these lights, passing Out of Town News. You will then approach another fork in the road — stay to the left of the fork and then immediately get into the right hand lane. This will take you onto Massachusetts Avenue, headed north. Go approximately 0.2 tenths of a mile past the Harvard Law School to Everett Street on your right. Turn right onto Everett Street and go to the next set of lights at Oxford Street. The building directly in front of you and beyond the lights is the Northwest Building. Behind that building is 38 Oxford Street.

**From the South**
Take I-93/Route 3 (South East Expressway) north and follow the above directions for Storrow Drive West.

**From the West**
Take I-90 (Massachusetts Turnpike) east. Get off the Cambridge/Allston exit (#18) and follow signs for Cambridge bearing to the right. As you come off the Turnpike exit, a 10-story hotel is on your right. Begin to move to the left lane as you pass the hotel. There is a traffic light at the corner of the hotel. Go through this light, but be prepared to make a left, before the bridge (do not cross over the bridge). You will now be on an entrance ramp leading onto Storrow Drive with very rapid traffic. Follow Storrow Drive approximately 1/2 mile. Stay in the right lane, which will become the “Harvard Square/Cambridge” exit on your right. Follow the directions given above for coming from the north starting with “. . . Take the exit and go over the Charles River . . .”

###PARKING

**On-street metered parking**
There are limited 2-hour metered spaces available on Oxford and Kirkland Streets (these meters only accept US quarters).

**Commercial Garages**
There are also commercial parking garages in Harvard Square, however the Division cannot validate parking for those businesses.

**University Day-Permit Parking**
Limited parking is available for department visitors in nearby University-operated parking garages. Day permits can be purchased up to 2 weeks in advance, which is highly recommended as the lots fill-up quickly, in person (at the Harvard Parking Office at 3 Bow Street) or [online](https://www2.uos.harvard.edu/cgi-bin/permit/purchase.pl). Daily permits are $14.00. All major credit cards are accepted, and all sales are final (no refunds or exchanges). If you wish to purchase a permit online, we recommend requesting a permit for the 52 Oxford Street Garage (Monday-Friday, 7:00am-11:30pm) or 10 Everett Street (Monday-Friday, 7:15am-11:00pm).
